### PR TITLE
Ironman/hardcore mode

### PR DESCRIPTION
**Pull request by AdamMil**
This change adds the beginning of an ironman/hardcore mode. In this first
iteration...
\* The intro conversation was modified to allow the selection of the
  game mode. (conversations.txt)
\* Autosaves were reduced and manual saves eliminated in challenge modes.
  (interfaces.txt, LoadPanel.cpp)
\+ The ability to hash save files (and other data files) was added. This is
  currently used to detect save file tampering (but not save scumming), and
  might eventually be used in some kind of online mod download/update thing
  that only downloads changed files or whatever. There's currently no penalty
  for editing the save file. (SHA1.\*, DataFile.\*, DataNode.\*, DataWriter.\*)
\* The player info was modified to expose the challenge mode and to add a
  condition when save file editing was detected. The RNG state is also loaded
  from the save file in challenge modes. (PlayerInfo.\*)
\+ The ability to save and load the RNG state was added. (Random.\*)
\+ A mission related to the challenge mode was added. It's quite lame at the
  moment. (challenge missions.txt)

### CODE REVIEW

File: EndlessSky.cbp [hunk 1]
```diff
@@ -77,6 +77,82 @@
 			<Add library="C:\Program Files\mingw64\x86_64-w64-mingw32\lib\libopengl32.a" />
 			<Add directory="C:/dev64/lib" />
 		</Linker>
+		<Unit filename="data/boarding missions.txt" />
```
Comment (AdamMil) [inline EndlessSky.cbp]: I found it convenient to have these here. I can remove them.

File: data/challenge missions.txt [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+		not saveFileTampered
+		or
+			year > 3014
+			and
+				year = 3014
+				month > 2
+			and
+				year = 3014
+				month = 2
+				day >= 17
```
Comment (AdamMil) [inline data/challenge missions.txt]: I could have used an event to encode this date comparison, but it seemed like events were for more important things...

**Comment by Bladewood** [comment]
what

File: data/challenge missions.txt [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+# Endless Sky is free software: you can redistribute it and/or modify it under the
+# terms of the GNU General Public License as published by the Free Software
+# Foundation, either version 3 of the License, or (at your option) any later version.
+#
+# Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+# PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+
+
+
```
Comment (AdamMil) [inline data/challenge missions.txt]: I wish these three missions could be combined into a single mission with different outcomes, but that would require branching inside of a trigger/MissionAction...

File: data/challenge missions.txt [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+		payment 100000
+		conversation
+			`	You return to the shipyard where he works. After a bit of time spent looking - half of it at all the ships you'll someday be able to afford - you see him standing around chatting. Doesn't he ever do anything useful? You walk over and catch his eye.`
+			`	"Wow, you're still alive. I gotta say, I didn't think you could do it!"`
+			`	"Yep, I won the bet. I think you owe me <payment>."`
+			`	"Yeah, yeah." He taps on his datapad for a moment and it spits out a credit chit, which he hands to you. "Hey... do you want to get a drink before you go?"`
+			`	You transfer the credits and hand it back to him. "Nope. Later!"`
+
+
+
```
Comment (AdamMil) [inline data/challenge missions.txt]: I wanted the "cheater" mission to be somewhat punishing, but I found that hard to implement.

File: data/challenge missions.txt [hunk 5]
```diff
@@ -0,0 +1,122 @@
+# Copyright (c) 2014 by Michael Zahniser
```
Comment (AdamMil) [inline data/challenge missions.txt]: These missions are lame. Suggestions wanted.

File: data/conversations.txt [hunk 6]
```diff
@@ -17,13 +17,45 @@ conversation "intro"
...
+	label end
+	`	He goes back to chatting with a group of guys. Finally. You return your gaze to the ships above. There are three models within your price range. Which one you choose will determine your future...`
+	`	So what'll it be?`
+		decline
 
 conversation "flight check: no energy"
 	scene "scene/engine2"
 	`You've heard enough horror stories about the sort of accidents that happen when starship systems malfunction to be completely convinced of the necessity of a pre-flight safety check. As you listen to the complex symphony of hums, rattles, and clicks of "<ship>" warming up, you run through a mental checklist. Computers: check. Life support: check.`
-	`	It is all going smoothly until it comes time to fire up your systems for a test run... and you discover that your generator does not produce enough power to power them. You'll have to either install some batteries, or switch to a more powerful generator.`	
-
```
Comment (AdamMil) [inline data/conversations.txt]: Just removing some unnecessary whitespace.

File: EndlessSky.cbp (hunk 1 above)
Comment (tehhowch) [inline EndlessSky.cbp]: Compiling the game does not depend on these text files at all, so they should not be present

File: data/challenge missions.txt [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+			`Looking over your calendar, you realize that you've survived your first three months. You should return to New Boston to collect on the bet you made with your old 'friend'.`
+			choice
+				`	Yeah, that's right! Time to get paid.`
+					accept
+				`	Nah, forget that guy. The money isn't worth having to talk to him again.`
+					decline
+	on complete
+		payment 50000
+		conversation
+			`	You return to the shipyard where he works. After a bit of time spent looking - half of it at all the ships you can't afford - you see him standing around chatting. Doesn't he ever do anything useful? You walk over and catch his eye.`
```
Comment (tehhowch) [inline data/challenge missions.txt]: The ships on New Boston aren't exactly expensive or desirable after 3 months.

File: data/challenge missions.txt [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+			`	"Wow, you're still alive. That's a surprise..."`
+			`	"Yep, I won the bet. Now about that <payment>..."`
+			`	"Yeah, yeah." He taps on his datapad for a moment and it spits out a credit chit, which he grudgingly hands to you. "Here."`
```
Comment (tehhowch) [inline data/challenge missions.txt]: chip

**Comment by Hadron1776** [comment]
I'm not sure I like things that actively restrict the player's options. If the player decides to do the bronze/ironman mode in your PR, and changes his mind (Or hits the choice by accident), he's screwed.

**Comment by AdamMil** [comment]
@Hadron1776 The only restriction is around saving and loading, which is standard for ironman. The choice is made before the game starts, and the effects are explained in the intro conversation. If the player doesn't want their saving and loading options restricted, they don't have to choose the challenge mode. If they choose it accidentally, they've lost nothing because the game hasn't started yet, and they can go back and choose something else.

File: source/DataWriter.cpp [hunk 9]
```diff
@@ -75,16 +77,19 @@ void DataWriter::Write()
...
 }
 
 
 
 // Decrease the indentation level.
 void DataWriter::EndChild()
 {
-	indent.erase(indent.length() - 1);
+	indent.erase(indent.length() - 1);
+	nextIndentLevel--;
```
Comment (tehhowch) [inline source/DataWriter.cpp]: --nextIndentLevel;

File: source/DataWriter.cpp [hunk 10]
```diff
@@ -75,16 +77,19 @@ void DataWriter::Write()
 
 // Increase the indentation level.
 void DataWriter::BeginChild()
-{
-	indent += '\t';
+{
+	// we compute 'indent' here rather than in EnsureIndent so it's correct for comments
+	indent += '\t';
+	nextIndentLevel++;
```
Comment (tehhowch) [inline source/DataWriter.cpp]: ++nextIndentLevel;

File: source/DataWriter.cpp [hunk 11]
```diff
@@ -129,4 +138,18 @@ void DataWriter::WriteToken(const char *a)
...
+
+
+
+// Ensure than the indentation level is set and hashed. We do this here rather
+// than immediately when BeginChild or EndChild is called so that code that
+// does Begin, Write, End, Begin, Write, End produces the same output and hash
+// as code that does Begin, Write, Write, End.
+void DataWriter::EnsureIndent()
+{
+	for(; nextIndentLevel > indentLevel; indentLevel++)
```
Comment (tehhowch) [inline source/DataWriter.cpp]: prefix instead of postfix

File: source/DataWriter.h [hunk 12]
```diff
@@ -14,7 +14,8 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 #define DATA_WRITER_H_
 
 #include <string>
-#include <sstream>
+#include <sstream>
+#include "SHA1.h"
```
Comment (tehhowch) [inline source/DataWriter.h]: file includes above `<>` includes

File: source/PlayerInfo.cpp [hunk 13]
```diff
@@ -104,6 +104,8 @@ void PlayerInfo::Load(const string &path)
 	filePath = path;
 	DataFile file(path);
 	
+	const DataNode *pHashNode = nullptr; // a pointer to the "hash" node, if any
```
Comment (tehhowch) [inline source/PlayerInfo.cpp]: No inline comments

File: source/SHA1.h [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#ifndef SHA1_H_
+#define SHA1_H_
+
+#include <stdint.h>
+#include <vector>
```
Comment (tehhowch) [inline source/SHA1.h]: alphabetized

File: data/challenge missions.txt [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+					accept
+				`	Nah, forget that guy. The money isn't worth seeing his face again.`
+					decline
+	on complete
+		payment 100000
+		conversation
+			`	You return to the shipyard where he works. After a bit of time spent looking - half of it at all the ships you'll someday be able to afford - you see him standing around chatting. Doesn't he ever do anything useful? You walk over and catch his eye.`
+			`	"Wow, you're still alive. I gotta say, I didn't think you could do it!"`
+			`	"Yep, I won the bet. I think you owe me <payment>."`
+			`	"Yeah, yeah." He taps on his datapad for a moment and it spits out a credit chit, which he hands to you. "Hey... do you want to get a drink before you go?"`
```
Comment (toilethinges) [inline data/challenge missions.txt]: chit -> chip again

Comment (AdamMil) [inline data/challenge missions.txt]: Changed in all three places.

**Comment by AdamMil** [comment]
Seems CodeBlocks was messing up the line endings. Fixed.

File: source/Random.h [hunk 16]
```diff
@@ -32,6 +32,15 @@ class Random {
...
 	
+	// Functions for getting random numbers that are "just for effect", meaning
+	// that calling them doesn't change the RNG used for more important game
+	// state. Examples are random pitch changes in sound effects and the
+	// generation of the background star field at program startup.
+	static uint32_t EffectInt();
+	static uint32_t EffectInt(uint32_t modulus);
+
+	static double EffectReal();
+
```
Comment (AdamMil) [inline source/Random.h]: I wish I could think of a better name for these. Ideally the RNG wouldn't be a static class and then it'd be easier to control changes to the game state, but that would be a more extensive change.

**Comment by endless-sky** [comment]
This has come up in the past. Players can choose to self-impose whatever sort of challenges they want (e.g. permadeath, upgrade ship only by plundering, etc.). Coding those restrictions into the game itself is just an unnecessary increase in complexity.

**Comment by KiLEdEnNis** [comment]
maybe eventually it'll be worth to add the option, but for now it's really unimportant imo

**Comment by AdamMil** [comment]
Certainly players cannot currently prevent the RNG from giving different results on a reload by force of will.

**Comment by endless-sky** [comment]
If you reload the game, you're not hardcore enough. Real hardcore players start a new pilot if they die.

**Comment by gunqqer** [comment]
Nah, true hardcore people do a full 32 zero out of their harddrive and reinstall everything everytime they die, using OS X but having a VM to use Linux which you have a VM inside of with Windows  in which you compile the game from source after running 10 mile on legos every time they die.

**Comment by Bladewood** [comment]
True hardcore people die in the real world and start a new life.

*leaves*

**Comment by gunqqer** [comment]
True, True hardcore people destroy the entire universe and start from a new big bang waiting the billions and billions of years until Earth becomes a thing again the whole time reflecting on how they fucked up.

**Comment by KiLEdEnNis** [comment]
true hardcore people feed off of the misery of the community when they tease us with new content in a screenshot in a random issue

**Comment by gunqqer** [comment]
Or don't even give us a screenshot, only a tid bit of text...
